spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ArrowVectorAccessors.java

```diff
@@ -0,0 +1,495 @@
+      return vector.get(rowId);
+    }
+  }
+
+  private static class DictionaryFloatAccessor extends DictionaryArrowVectorAccessor {
+
+    private final IntVector vector;
+
+    DictionaryFloatAccessor(IntVector vector, Dictionary dictionary) {
+      super(vector, dictionary);
+      this.vector = vector;
+    }
+
+    @Override
+    final float getFloat(int rowId) {
+      return parquetDictionary.decodeToFloat(vector.get(rowId));
+    }
+  }
+
+  private static class DoubleAccessor extends ArrowVectorAccessor {
+
+    private final Float8Vector vector;
+
+    DoubleAccessor(Float8Vector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final double getDouble(int rowId) {
+      return vector.get(rowId);
+    }
+  }
+
+  private static class DictionaryDoubleAccessor extends DictionaryArrowVectorAccessor {
+
+    private final IntVector vector;
+
+    DictionaryDoubleAccessor(IntVector vector, Dictionary dictionary) {
+      super(vector, dictionary);
+      this.vector = vector;
+    }
+
+    @Override
+    final double getDouble(int rowId) {
+      return parquetDictionary.decodeToDouble(vector.get(rowId));
+    }
+  }
+
+  private static class DecimalAccessor extends ArrowVectorAccessor {
+
+    private final IcebergArrowVectors.DecimalArrowVector vector;
+
+    DecimalAccessor(IcebergArrowVectors.DecimalArrowVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final Decimal getDecimal(int rowId, int precision, int scale) {
+      return Decimal.apply(vector.getObject(rowId), precision, scale);
+    }
+  }
+
+  private static class StringAccessor extends ArrowVectorAccessor {
+
+    private final IcebergArrowVectors.VarcharArrowVector vector;
+    private final NullableVarCharHolder stringResult = new NullableVarCharHolder();
+
+    StringAccessor(IcebergArrowVectors.VarcharArrowVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final UTF8String getUTF8String(int rowId) {
+      vector.get(rowId, stringResult);
+      if (stringResult.isSet == 0) {
+        return null;
+      } else {
+        return UTF8String.fromAddress(
+            null,
+            stringResult.buffer.memoryAddress() + stringResult.start,
+            stringResult.end - stringResult.start);
+      }
+    }
+  }
+
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  private abstract static class DictionaryArrowVectorAccessor extends ArrowVectorAccessor {
+    final Dictionary parquetDictionary;
+    final IntVector dictionaryVector;
+
+    private DictionaryArrowVectorAccessor(IntVector vector, Dictionary dictionary) {
+      super(vector);
+      this.dictionaryVector = vector;
+      this.parquetDictionary = dictionary;
+    }
+  }
+
+  private static class DictionaryStringAccessor extends DictionaryArrowVectorAccessor {
+
+    DictionaryStringAccessor(IntVector vector, Dictionary dictionary) {
+      super(vector, dictionary);
```

Comment:
I think that the dictionary should be eagerly decoded, for a few reasons:
1. `getBytesUnsafe` will make a copy of the bytes when the binary is backed by a `ByteBuffer` or an array slice. It may also cache values and perform checks.
2. `getBytesUnsafe` should only be used when the bytes are immediately consumed, but `UTF8String` keeps a reference, so even if there isn't a copy this could result in a correctness problem
3. `decodeToBinary` should incur a dynamic dispatch cost (this also applies to the other dictionary readers)

It's fairly easy to eagerly decode:
```java
  private static class DictionaryStringAccessor extends DictionaryArrowVectorAccessor {
    private IntVector offsetVector;
    private UTF8String[] decodedDictionary;

    DictionaryStringAccessor(IntVector vector, Dictionary dictionary) {
      super(vector, dictionary);
      this.offsetVector = vector;
      this.decodedDictionary = IntStream.range(0, dictionary.getMaxId() + 1)
          .mapToObj(dictionary::decodeToBinary)
          .map(binary -> UTF8String.fromBytes(binary.getBytes()))
          .toArray(UTF8String[]::new);
    }

    @Override
    final UTF8String getUTF8String(int rowId) {
      int offset = offsetVector.get(rowId);
      return decodedDictionary[offset];
    }
  }
```